projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/c/LLVMSignal.java

```diff
@@ -0,0 +1,425 @@
+
+    /**
+     * Registers a signal handler using sun.misc.SignalHandler. Unfortunately, using signals in java
+     * leads to some problems which are not resolved in our implementation yet.
+     *
+     * One of this issue is, that signals are executed in an asynchronous way, which means raise()
+     * exits before the signal was handled. Another Issue is that Java already registered some
+     * signal handlers, which therefore cannot be used in sulong.
+     *
+     * Therefore, our implementation does not comply with the ANSI C standard and could lead to
+     * timing issues when calling multiple signals in a defined sequence, or when a program has to
+     * wait until the signal was handled (which is not guaranteed because of the asynchronous
+     * behavior in our implementation).
+     */
+    private static final class LLVMSignalHandler implements SignalHandler, LLVMThreadNode {
+
+        private final Signal signal;
+        private final LLVMFunctionDescriptor function;
+        private final LLVMContext context;
+        private RootCallTarget callTarget;
+        private final LLVMStack stack = new LLVMStack();
+
+        private final Lock lock = new ReentrantLock();
+        private final AtomicBoolean isRunning = new AtomicBoolean(false);
+
+        @TruffleBoundary
+        private LLVMSignalHandler(Signal signal, LLVMFunctionDescriptor function) throws IllegalArgumentException {
+            this.signal = signal;
+            this.function = function;
+            LLVMFunctionNode functionNode = LLVMFunctionLiteralNodeGen.create(function);
+
+            LLVMAddressLiteralNode signalStack = new LLVMAddressLiteralNode(stack.allocate(TMP_SIGNAL_STACK_SIZE_BYTE));
+            LLVMI32LiteralNode sigNumArg = new LLVMI32LiteralNode(signal.getNumber());
+            LLVMExpressionNode[] args = {signalStack, sigNumArg};
+
+            context = LLVMLanguage.INSTANCE.findContext0(LLVMLanguage.INSTANCE.createFindContextNode0());
+
+            LLVMUnresolvedCallNode callNode = new LLVMUnresolvedCallNode(functionNode, args, LLVMRuntimeType.VOID, context);
+
+            callTarget = Truffle.getRuntime().createCallTarget(
+                            new LLVMFunctionStartNode(callNode,
+                                            new LLVMNode[]{},
+                                            new LLVMNode[]{},
+                                            SourceSection.createUnavailable("", null),
+                                            new FrameDescriptor(), ""));
+
+            lock.lock();
+            try {
+                if (function.equals(LLVM_SIG_DFL)) {
+                    Signal.handle(signal, SignalHandler.SIG_DFL);
+                    stack.free();
+                } else if (function.equals(LLVM_SIG_IGN)) {
+                    Signal.handle(signal, SignalHandler.SIG_IGN);
+                    stack.free();
+                } else {
+                    Signal.handle(signal, this);
```

Comment:
Is it possible that you just allocate the stack here, if you free it immediately in the other cases?